### ISSUE TITLE
slight variation of dynamics

### ISSUE THREAD

**Issue by RustoMCSpit** [nicolasbrailo/PianOli#98]
every time a note is played there should be a very, very tiny variation of the dynamics to make it sound more human

**Comment by juleskers** [comment]
Oof, I had to look up what "[dynamics](https://en.wikipedia.org/wiki/Dynamics_(music))" meant (my last experience with musical theory was a few weeks of long-since-forgotten pianolessons, and a middle school music teacher who never showed up..)

Do I understand correctly: slight volume changes for each note?
The android API allows some control of playback volume, per note-playing-event.

The piece of code that actually triggers playing is here: https://github.com/nicolasbrailo/PianOli/blob/7557a291f185b9ac9629fb47ec3cb784b84a808c/app/src/main/java/com/nicobrailo/pianoli/sound/SampledSoundSet.java#L121

We use the Android [`SoundPool.play(,,)`-API](https://developer.android.com/reference/android/media/SoundPool#play(int,%20float,%20float,%20int,%20int,%20float)), to give you some indication as to what is easily achievable from the technical side (of note: parameters `leftVolume`, `rightVolume`, `loop`-count, `rate`).

**Comment by RustoMCSpit** [comment]
> Do I understand correctly: slight volume changes for each note?

yeah! having it randomly be slightly louder or softer makes it a bit more human and less robotic

**Comment by nicolasbrailo** [comment]
I don't think we could achieve a close effect by using the volume control APIs, to come close we'd need a filter to control the timbre, and a compressor we can use to change attack and release times. Would be an awesome change, but we'd need to rewrite the entire audio playback "backend" (there is no "backend" now, actually, we'd need to build one :))

@RustoMCSpit thank you for all your suggestions, I think they are awesome and it'll probably take me a while to go through them and see how feasible they are, but I think you are looking for a different app. Pianoli is not *really* a music app, it's a game for children, that happens to have sound playing capabilities in the shape of a piano.

A music app similar to this app would be built on top of a sound engine or DAW, something that can be used to process [real time] audio. Pianoli doesn't have any of those capabilities, it can only read mp3 files and play them. It would be awesome to rebuild the app on top of an audio engine, but that'd be Pianoli V2 pretty much from scratch. I'd be delighted to help get PRs landed, but I don't think I'd work on that myself any time soon.

**Comment by RustoMCSpit** [comment]
theyre just suggestions dont worry, do whatever you want. thnak you for your kindness

**Comment by nicolasbrailo** [comment]
They are great suggestions, and I'd love to see some of them as part of the
app some day, but I prefer to be honest: it's not likely that anyone will
find the time required to implement an engine that can support them. (Now
prove me wrong, Internet!)
